checks/http.go

```diff
@@ -0,0 +1,242 @@
+	p := RaintankProbeHTTP{}
+	err := json.Unmarshal(body, &p)
+	if err != nil {
+		log.Fatalf("failed to parse settings. %v", err.Error())
+		return nil, err
+	}
+	if port, err := strconv.ParseInt(p.Port, 10, 32); err != nil || port < 1 || port > 65535 {
+		log.Fatal("failed to parse settings. Invalid port")
+		return nil, err
+	}
+	if !strings.HasPrefix(p.Path, "/") {
+		return nil, fmt.Errorf("invalid path. must start with /")
+	}
+	return &p, nil
+}
+
+// Results interface
+func (p *RaintankProbeHTTP) Results() interface{} {
+	return p.Result
+}
+
+
+// Run checking
+func (p *RaintankProbeHTTP) Run() error {
+	p.Result = &HTTPResult{}
+	
+	if p.Method == "" {
+		p.Method = "GET"
+	}
+	
+	// reader
+	url := fmt.Sprintf("http://%s:%s%s", p.Host, p.Port, p.Path)
+	request, err := http.NewRequest(p.Method, url, nil)
+	request.Header.Set("Connection", "close")
+	
+	// Parsing header (use fake request)
+	if p.Headers != "" {
+		headReader := bufio.NewReader(strings.NewReader("GET / HTTP/1.1\r\n" + p.Headers + "\r\n\r\n"))
+		dummyRequest, err := http.ReadRequest(headReader)
+		if err != nil {
+			msg := err.Error()//"failed to parse header."
+			p.Result.Error = &msg
+			return nil
+		}
+		
+		for key := range dummyRequest.Header {
+			request.Header.Set(key, dummyRequest.Header.Get(key))
+		}
+	}
+	
+	if err != nil {
+		msg := "connection closed"
+		p.Result.Error = &msg
+		return nil
+	}
+	
+	// Dialing
+	start := time.Now()
+	conn, err := net.Dial("tcp", p.Host+":"+p.Port)
+	if err != nil {
+		msg := err.Error()
+		p.Result.Error = &msg
+		return nil
+	}
+	duration := time.Since(start).Seconds() * 1000
+	p.Result.Connect = &duration
+	
+	// DNS
+	step := time.Now()
+	addrs, err := net.LookupHost(p.Host)
+	if err != nil || len(addrs) < 1 {
+		msg := "failed to resolve hostname to IP."
+		p.Result.Error = &msg
+		return nil
+	}
+	duration = time.Since(step).Seconds() * 1000
+	p.Result.DNS = &duration
+
+	// Send
+	step = time.Now()
+	if err := request.Write(conn); err != nil {
+		msg := err.Error()
+		p.Result.Error = &msg
+		return nil
+	}
+	send := time.Since(step).Seconds() * 1000
+	p.Result.Send = &send
+	defer conn.Close()
+	
+	// Wait & Receive
+	step = time.Now()
+	data := make([]byte, 1)
+	// read first data
+	_, err = conn.Read(data)
+	wait := time.Since(step).Seconds() * 1000
+	p.Result.Wait = &wait
+	
+	if err != nil {
+		if err != io.EOF {
+			msg := "Read error"
+			p.Result.Error = &msg
+			return nil
+		}
+	}
+
+	var buf bytes.Buffer
+	
+	buf.Write(data)
+	io.CopyBuffer(&buf, conn, data)
+	defer conn.Close()
+	
+	recv := time.Since(step).Seconds() * 1000
+	/* 
+		Total time 
+	*/
+	total := time.Since(start).Seconds() * 1000
+	p.Result.Total = &total
+	
+	p.Result.Recv = &recv
+	defer conn.Close()
+	
+	readbuffer := bytes.NewBuffer(buf.Bytes())
+	response, err := http.ReadResponse(bufio.NewReader(readbuffer), request)
+	
+	// Error response
+	statusCode := float64(response.StatusCode)
+	if statusCode >= 400 {
```

Comment:
This needs to be moved to run just before the regex check.
Even if statusCode is >400 we still want to collect all of the timing data.